executor/adapter.go

```diff
@@ -141,8 +142,40 @@ func (a *statementAdapter) Exec(ctx context.Context) (rset.Recordset, error) {
 	if b.err != nil {
 		return nil, errors.Trace(b.err)
 	}
-	fields := make([]*field.ResultField, 0, len(e.Fields()))
-	for _, v := range e.Fields() {
+
+	if executorExec, ok := e.(*ExecuteExec); ok {
+		err := executorExec.Build()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if executorExec.OldStmt != nil {
+			return executorExec.OldStmt.Exec(executorExec.Ctx)
+		}
+		e = executorExec.StmtExec
+	}
+
+	if len(e.Fields()) == 0 {
+		// No result fields means no Recordset.
+		for {
+			defer e.Close()
```

Comment:
is it safe to call defer in for? 
I knew sometimes defer only be called when function returns.

@qiuyesuifeng